[to_edge] Allow core aten op exception list

larryliu0820 opened the pull request:
Summary: Currently when a non-core ATen operator shows up in the exported graph, `to_edge()` will fail and the only option is to disable IR validity check by setting `_check_ir_validity=False`. However this is unsafe to do, instead we should still run the rest of the checks.

This PR adds support to allow users to bypass core ATen ops check, by passing a list of non-core ATen ops into `to_edge()`.

Note that:

* This is different than `ops_set_to_not_decompose` in `to_edge_transform_and_lower`, as the ops in `_core_aten_ops_exception_list` are not intended to be kept but more likely showing up because of missing decompositions or missing core ATen tag in `native_functions.yaml`. For this reason, we are combining two lists (`ops_set_to_not_decompose` and `_core_aten_ops_exception_list`) and pass to verifier.
* I updated the error log to encourage people to use `_core_aten_ops_exception_list` instead of using `_check_ir_validity=False`.

Test Plan: Added unit test

Reviewers:

Subscribers:

Tasks:

Tags:

JacobSzwejbka commented:
> as the ops in _core_aten_ops_exception_list are not intended to be kept but more likely showing up because of missing decompositions or missing core ATen tag in native_functions.yaml.

Why dont we just fix these when they happen then instead of allowing people to bypass?

larryliu0820 commented:
> > as the ops in _core_aten_ops_exception_list are not intended to be kept but more likely showing up because of missing decompositions or missing core ATen tag in native_functions.yaml.
> 
> Why dont we just fix these when they happen then instead of allowing people to bypass?
> 
> Fundamentally Im unsure of the value of a verifier if the recommended course of action when it fails is just to bypass. If bypassing it is fine then why not just always bypass? If its not fine in general why is it fine for this specific user who its failing for?

It takes time for us to implement decompositions or adding an operator into core ATen. Bypassing core ATen op check doesn't mean we are bypassing all other checks in the verifier, it's an escape patch for a user to express "I know this op exists in the graph that is not core ATen and there's no portable kernel for it. I'm confident it will be handled downstream".

larryliu0820 committed:
Allow core aten op exception list (#5237)

Summary:
Currently when a non-core ATen operator shows up in the exported graph, `to_edge()` will fail and the only option is to disable IR validity check by setting `_check_ir_validity=False`. However this is unsafe to do, instead we should still run the rest of the checks.

This PR adds support to allow users to bypass core ATen ops check, by passing a list of non-core ATen ops into `to_edge()`.

Note that:

* This is different than `ops_set_to_not_decompose` in `to_edge_transform_and_lower`, as the ops in `_core_aten_ops_exception_list` are not intended to be kept but more likely showing up because of missing decompositions or missing core ATen tag in `native_functions.yaml`. For this reason, we are combining two lists (`ops_set_to_not_decompose` and `_core_aten_ops_exception_list`) and pass to verifier.
* I updated the error log to encourage people to use `_core_aten_ops_exception_list` instead of using `_check_ir_validity=False`.


Test Plan: Added unit test

Reviewed By: tarun292

Differential Revision: D62469015

Pulled By: larryliu0820
